stumpy/stump.py

```diff
@@ -182,13 +182,35 @@ def _compute_diagonal(
 
             if T_B_subseq_isfinite[i + k] and T_A_subseq_isfinite[i]:
                 # Neither subsequence contains NaNs
-                if T_B_subseq_isconstant[i + k] or T_A_subseq_isconstant[i]:
-                    pearson = 0.5
-                else:
-                    pearson = cov * Σ_T_inverse[i + k] * σ_Q_inverse[i]
-
                 if T_B_subseq_isconstant[i + k] and T_A_subseq_isconstant[i]:
                     pearson = 1.0
+                elif T_B_subseq_isconstant[i + k] or T_A_subseq_isconstant[i]:
+                    pearson = 0.5
+                else:
+                    denom = Σ_T[i + k] * σ_Q[i]
```

Comment:
This:

```
                    denom = Σ_T[i + k] * σ_Q[i]
                    if denom < config.STUMPY_MIN_STD_AB:
                        cov = (
                            np.dot(
                                (T_B[i + k : i + k + m] - M_T[i + k]),
                                (T_A[i : i + m] - μ_Q[i]),
                            )
                            * m_inverse
                        )
```

Should not be here. I think we want to move this higher up:

```
        for i in iter_range:
            denom = Σ_T[i + k] * σ_Q[i]
            if i == 0 or (k < 0 and i == -k) or denom < config.STUMPY_MIN_STDDEVS:
                cov = (
                    np.dot(
                        (T_B[i + k : i + k + m] - M_T[i + k]), (T_A[i : i + m] - μ_Q[i])
                    )
                    * m_inverse
                )
            else:
                # The next lines are equivalent and left for reference
                # cov = cov + constant * (
                #     (T_B[i + k + m - 1] - M_T_m_1[i + k])
                #     * (T_A[i + m - 1] - μ_Q_m_1[i])
                #     - (T_B[i + k - 1] - M_T_m_1[i + k]) * (T_A[i - 1] - μ_Q_m_1[i])
                # )
                cov = cov + constant * (
                    cov_a[i + k] * cov_b[i] - cov_c[i + k] * cov_d[i]
                )
```

This would make things much cleaner and without repeating the `np.dot` code in multiple places. Note that the inverse stddevs were used because they would avoid the slightly more costly division operation (compared with simply multiplication operations). By using the inverse, the inverse (division) of each subsequence's stddev is only computed once and then used in multiplications. Here, we are basically re-computing the division every single time each subsequence's stddev is compared with another. There is no reason why you couldn't do something like:

```
stddevs_inverse = Σ_T_inverse[i + k] * σ_Q_inverse[i]
if i == 0 or (k < 0 and i == -k)  or stddevs_inverse > config.STUMPY_MIN_STDDEVS_INVERSE:
```

Reply:
> This would make things much cleaner and without repeating the `np.dot` code in multiple places

Yeah...right! Thanks for the feedback!



> Note that the inverse stddevs were used because they would avoid the slightly more costly division operation (compared with simply multiplication operations). By using the inverse, the inverse (division) of each subsequence's stddev is only computed once and then used in multiplications. Here, we are basically re-computing the division every single time each subsequence's stddev is compared with another. There is no reason why you couldn't do something like:

Yeah...I remember your explanation. But, I thought it is not possible to use inverse here. I was wrong! `config.STUMPY_MIN_STDDEVS_INVERSE = 1.0` (default) should work.